Documentation/contributing.rst

```diff
@@ -431,11 +432,19 @@ Once you add the libraries that you need you can save the file and run
 
 .. code:: bash
 
-    $ dep ensure -v
+    $ dep ensure -v -add github.com/containernetworking/cni
```

Comment:
should be kept as `$ dep ensure -v`. Modifying the `Gopkg.toml` does not require to run `dep ensure` with `-add`

Reply:
In my experience, running without the `-add <package>` or `-update <package>` will try to resolve all the vendor dependencies. Sometimes you want to do that, but many times I'm out of my depth trying to figure out the myriad of failures being reported on dependencies I know nothing about. With the specified dependency, `dep` will only try to resolve that specific dependency, which is much more likely to succeed.

So, while, strictly speaking, `-add` or `-update` is not *necessary*, using them makes the process much less painful, and many times allows me to complete the task at hand rather than try to fix up `dep` failures that also keep changing from a run to the next.

Reply:
>With the specified dependency, dep will only try to resolve that specific dependency, which is much more likely to succeed.

I know, but it also means the next developer that will run `dep ensure` will have all of the myriad of failures to deal with. For this reason, `dep ensure` should be used locally. This will resolve all indirect dependencies that have to be dealt by updating / adding a specific direct dependency

Reply:
Do you mean that `-add` or `-update` do not resolve transitive dependencies? Also, what does it mean to use `dep ensure` *locally*?

Reply:
From `dep -examples`:
```
dep ensure -update

    Update all dependencies to the latest versions allowed by Gopkg.toml,
    ignoring any versions recorded in Gopkg.lock. Update the lock file with any
    changes. (NOTE: Not recommended. Updating one/some dependencies at a time is
    preferred.)
```

Note that it recommends updating one dependency at a time, rather than trying to ensure all of them at once.

Reply:
that's my understanding from the help function of `dep ensure`
```
  -add          add new dependencies, or populate Gopkg.toml with constraints for existing dependencies (default: false)
  -dry-run      only report the changes that would be made (default: false)
  -examples     print detailed usage examples (default: false)
  -no-vendor    update Gopkg.lock (if needed), but do not update vendor/ (default: false)
  -update       update the named dependencies (or all, if none are named) in Gopkg.lock to the latest allowed by Gopkg.toml (default: false)
  -v            enable verbose logging (default: false)
  -vendor-only  populate vendor/ from Gopkg.lock without updating it first (default: false)
```
>Also, what does it mean to use dep ensure locally?

run it in a laptop.

Reply:
IMO `dep ensure -update` *must* resolve transitive dependencies, otherwise your project would not compile. From the FAQ:
```
Transitive dependencies are the dependencies of your dependencies.
Necessary to compile but are not directly used by your code.
```

So far I've always run `dep` in a laptop, but in the Cilium dev VM. Do you mean that I should run it in the Mac instead?

Reply:
I agree with @jrajahalme . We should be following what the documentation says for `dep ensure`, specifically: 

> Note that it recommends updating one dependency at a time, rather than trying to ensure all of them at once.

Reply:
Hmm, I almost never had a problem running `dep ensure -update` (with a stable version of `dep`) w/o specifying a dependency.

https://golang.github.io/dep/docs/daily-dep.html#updating-dependencies is less strict about using it:
> Ideally, updating a dependency project to a newer version is a single command:
>
> $ dep ensure -update github.com/foo/bar
>
> This also works without arguments to try to update all dependencies (though it's generally not   recommended):
>
> $ dep ensure -update

Reply:
@brb @ianvernon @jrajahalme I meant modifying or add a new dependency in `Gopkg.toml` and run `dep ensure -v` is enough. Running `dep ensure -update` / `dep ensure -update <dependency>` will update the dependency to the latest version(s) available which likely not what we want for existing dependencies. For both cases, adding an updating a new dependency, changing the `Gopkg.toml` and running `dep ensure -v` should be enough as this last command does, quoting from docs:

>Hey dep, please make sure that my project is in sync: that Gopkg.lock satisfies all the imports in my project, and all the rules in Gopkg.toml, and that vendor/ contains exactly what Gopkg.lock says it should."

I spoke with @jrajahalme privately and we our conclusion was an out-of-sync git repo that make him believe the issue was related with the way that he was adding / updating dependencies with dep.